<a href="https://colab.research.google.com/github/vishwanathbeena/appliedai_practice/blob/main/Assignment_3_Instructions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment

<font face='georgia'>
    
   <h4><strong>What does tf-idf mean?</strong></h4>

   <p>    
Tf-idf stands for <em>term frequency-inverse document frequency</em>, and the tf-idf weight is a weight often used in information retrieval and text mining. This weight is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The importance increases proportionally to the number of times a word appears in the document but is offset by the frequency of the word in the corpus. Variations of the tf-idf weighting scheme are often used by search engines as a central tool in scoring and ranking a document's relevance given a user query.
</p>
    
   <p>
One of the simplest ranking functions is computed by summing the tf-idf for each query term; many more sophisticated ranking functions are variants of this simple model.
</p>
    
   <p>
Tf-idf can be successfully used for stop-words filtering in various subject fields including text summarization and classification.
</p>
    
</font>

<font face='georgia'>
    <h4><strong>How to Compute:</strong></h4>

Typically, the tf-idf weight is composed by two terms: the first computes the normalized Term Frequency (TF), aka. the number of times a word appears in a document, divided by the total number of words in that document; the second term is the Inverse Document Frequency (IDF), computed as the logarithm of the number of the documents in the corpus divided by the number of documents where the specific term appears.

 <ul>
    <li>
<strong>TF:</strong> Term Frequency, which measures how frequently a term occurs in a document. Since every document is different in length, it is possible that a term would appear much more times in long documents than shorter ones. Thus, the term frequency is often divided by the document length (aka. the total number of terms in the document) as a way of normalization: <br>

$TF(t) = \frac{\text{Number of times term t appears in a document}}{\text{Total number of terms in the document}}.$
</li>
<li>
<strong>IDF:</strong> Inverse Document Frequency, which measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones, by computing the following: <br>

$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}}.$
for numerical stabiltiy we will be changing this formula little bit
$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}+1}.$
</li>
</ul>

<br>
<h4><strong>Example</strong></h4>
<p>

Consider a document containing 100 words wherein the word cat appears 3 times. The term frequency (i.e., tf) for cat is then (3 / 100) = 0.03. Now, assume we have 10 million documents and the word cat appears in one thousand of these. Then, the inverse document frequency (i.e., idf) is calculated as log(10,000,000 / 1,000) = 4. Thus, the Tf-idf weight is the product of these quantities: 0.03 * 4 = 0.12.
</p>
</font>

## Task-1

<font face='georgia'>
    <h4><strong>1. Build a TFIDF Vectorizer & compare its results with Sklearn:</strong></h4>

<ul>
    <li> As a part of this task you will be implementing TFIDF vectorizer on a collection of text documents.</li>
    <br>
    <li> You should compare the results of your own implementation of TFIDF vectorizer with that of sklearns implemenation TFIDF vectorizer.</li>
    <br>
    <li> Sklearn does few more tweaks in the implementation of its version of TFIDF vectorizer, so to replicate the exact results you would need to add following things to your custom implementation of tfidf vectorizer:
       <ol>
        <li> Sklearn has its vocabulary generated from idf sroted in alphabetical order</li>
        <li> Sklearn formula of idf is different from the standard textbook formula. Here the constant <strong>"1"</strong> is added to the numerator and denominator of the idf as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions.
            
 $IDF(t) = 1+\log_{e}\frac{1\text{ }+\text{ Total  number of documents in collection}} {1+\text{Number of documents with term t in it}}.$
        </li>
        <li> Sklearn applies L2-normalization on its output matrix.</li>
        <li> The final output of sklearn tfidf vectorizer is a sparse matrix.</li>
    </ol>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer.</li>
        <li> Print out the alphabetically sorted voacb after you fit your data and check if its the same as that of the feature names from sklearn tfidf vectorizer. </li>
        <li> Print out the idf values from your implementation and check if its the same as that of sklearns tfidf vectorizer idf values. </li>
        <li> Once you get your voacb and idf values to be same as that of sklearns implementation of tfidf vectorizer, proceed to the below steps. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> After completing the above steps, print the output of your custom implementation and compare it with sklearns implementation of tfidf vectorizer.</li>
        <li> To check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it.</li>
        </ol>
    </li>
    <br>
   </ul>

  <p> <font color="#e60000"><strong>Note-1: </strong></font> All the necessary outputs of sklearns tfidf vectorizer have been provided as reference in this notebook, you can compare your outputs as mentioned in the above steps, with these outputs.<br>
   <font color="#e60000"><strong>Note-2: </strong></font> The output of your custom implementation and that of sklearns implementation would match only with the collection of document strings provided to you as reference in this notebook. It would not match for strings that contain capital letters or punctuations, etc, because sklearn version of tfidf vectorizer deals with such strings in a different way. To know further details about how sklearn tfidf vectorizer works with such string, you can always refer to its official documentation.<br>
   <font color="#e60000"><strong>Note-3: </strong></font> During this task, it would be helpful for you to debug the code you write with print statements wherever necessary. But when you are finally submitting the assignment, make sure your code is readable and try not to print things which are not part of this task.
    </p>

### Corpus

In [ ]:
## SkLearn# Collection of string documents

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]


### SkLearn Implementation

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
vectorizer.fit(corpus)
skl_output = vectorizer.transform(corpus)

In [ ]:
# sklearn feature names, they are sorted in alphabetic order by default.

print(vectorizer.get_feature_names())

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


In [ ]:
# Here we will print the sklearn tfidf vectorizer idf values after applying the fit method
# After using the fit function on the corpus the vocab has 9 words in it, and each has its idf value.

print(vectorizer.idf_)

[1.91629073 1.22314355 1.51082562 1.         1.91629073 1.91629073
 1.         1.91629073 1.        ]


In [ ]:
# shape of sklearn tfidf vectorizer output after applying transform method.

skl_output.shape

(4, 9)

In [ ]:
# sklearn tfidf values for first line of the above corpus.
# Here the output is a sparse matrix

print(skl_output[0])

  (0, 8)	0.38408524091481483
  (0, 6)	0.38408524091481483
  (0, 3)	0.38408524091481483
  (0, 2)	0.5802858236844359
  (0, 1)	0.46979138557992045


In [ ]:
# sklearn tfidf values for first line of the above corpus.
# To understand the output better, here we are converting the sparse output matrix to dense matrix and printing it.
# Notice that this output is normalized using L2 normalization. sklearn does this by default.

print(skl_output[0].toarray())

[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


### Your custom implementation

In [ ]:
# Write your code here.
# Make sure its well documented and readble with appropriate comments.
# Compare your results with the above sklearn tfidf vectorizer
# You are not supposed to use any other library apart from the ones given below

from collections import Counter
from tqdm import tqdm
from scipy.sparse import csr_matrix
import math
import operator
from sklearn.preprocessing import normalize
import numpy

def fit(dataset):

  if isinstance(dataset,list):
    idf_len = len(dataset) + 1
    idf_dict = {}
    for doc in dataset:#for every line in inptlist
      doc_dict = Counter(doc.split(" "))#create a counter of words in that line.
      for item in doc_dict:
        if item in idf_dict:#For every word in doc_dict Counter,We are adding it to idf_dict Counter and increasing count by 1,if it already exists,means word is found in new line too
          idf_dict[item] +=1
        else:
            idf_dict[item] = 1
    idf_final_dict = {}
    for idx,key in enumerate(sorted(idf_dict)):#sorting dictionary based on keys and creating a final dictionary with IDF value and index as value
      idf_value = round(1 + (math.log(idf_len/(1+idf_dict[key]))),8)
      idf_final_dict[key] = [idf_value,idx]
    return idf_final_dict    
  else:
    return 'need list as argument'
def transform(dataset,vocab):
    rows = []
    columns = []
    values = []
    for idx,doc in enumerate(dataset):
      doc_dict = Counter(doc.split(" "))
      for item in doc_dict:
          if item in vocab:
            rows.append(idx)
            columns.append(vocab[item][1])
            values.append(vocab[item][0]) 
    return normalize(csr_matrix((values, (rows,columns)), shape=(len(dataset),len(vocab))))
  





[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


# Printing feature names,IDF values and their respective column index in a dictionary
Key is the feature and first item in list i IDF value and second item is its position in feature bag.

In [ ]:
idf_dict = fit(corpus)
print(idf_dict)

{'and': [1.91629073, 0], 'document': [1.22314355, 1], 'first': [1.51082562, 2], 'is': [1.0, 3], 'one': [1.91629073, 4], 'second': [1.91629073, 5], 'the': [1.0, 6], 'third': [1.91629073, 7], 'this': [1.0, 8]}


# Tesing transform function and priting idf values of first document in corpus
It is matching with sklearn implementation given above,which is 
[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]

In [ ]:
transform_op = transform(corpus,idf_dict)
print(transform_op[0].toarray())

[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


## Task-2

<font face='georgia'>
    <h4><strong>2. Implement max features functionality:</strong></h4>

<ul>
    <li> As a part of this task you have to modify your fit and transform functions so that your vocab will contain only 50 terms with top idf scores.</li>
    <br>
    <li>This task is similar to your previous task, just that here your vocabulary is limited to only top 50 features names based on their idf values. Basically your output will have exactly 50 columns and the number of rows will depend on the number of documents you have in your corpus.</li>
    <br>
    <li>Here you will be give a pickle file, with file name <strong>cleaned_strings</strong>. You would have to load the corpus from this file and use it as input to your tfidf vectorizer.</li>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer, just like in the previous task. Additionally, here you have to limit the number of features generated to 50 as described above.</li>
        <li> Now sort your vocab based in descending order of idf values and print out the words in the sorted voacb after you fit your data. Here you should be getting only 50 terms in your vocab. And make sure to print idf values for each term in your vocab. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> Now check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it. And this dense matrix should contain 1 row and 50 columns. </li>
        </ol>
    </li>
    <br>
   </ul>

In [ ]:
# Below is the code to load the cleaned_strings pickle file provided
# Here corpus is of list type

import pickle
with open('/content/cleaned_strings', 'rb') as f:
    corpus = pickle.load(f)
print(corpus[0])
    
# printing the length of the corpus loaded
print("Number of documents in corpus = ",len(corpus))

slow moving aimless movie distressed drifting young man
Number of documents in corpus =  746


In [ ]:
# Write your code here.
# Try not to hardcode any values.
# Make sure its well documented and readble with appropriate comments.
from collections import Counter
from tqdm import tqdm
from scipy.sparse import csr_matrix
import math
import operator
from sklearn.preprocessing import normalize
import numpy
def fit_top_fifty(dataset):
  if isinstance(dataset,list):
    idf_len = len(dataset) + 1
    idf_dict = {}
    for doc in dataset:#for every line in inptlist
      doc_dict = Counter(doc.split(" "))#create a counter of words in that line.
      for item in doc_dict:
        if item in idf_dict:#For every word in doc_dict Counter,We are adding it to idf_dict Counter and increasing count by 1,if it already exists,means word is found in new line too
          idf_dict[item] +=1
        else:
            idf_dict[item] = 1
    idf_final_list = []
    for key in idf_dict:
      idf_value = round(1 + (math.log(idf_len/(1+idf_dict[key]))),8)
      idf_final_list.append([key,idf_value])
    idf_final_list = sorted(idf_final_list,key=(lambda x:x[1]),reverse=True)[0:50] #took first elements with highest IDF value
    idf_final_dict = {item[0]:[item[1],idx] for idx,item in enumerate(sorted(idf_final_list,key=(lambda x:x[0])))}#after getting top 50 elements, sorted them again based on feature name in alphabetica order
    return idf_final_dict    
  else:
    return 'need list as argument'

def transform_top_fifty(dataset,vocab):
    rows = []
    columns = []
    values = []
    for idx,doc in enumerate(dataset):
      doc_dict = Counter(doc.split(" "))
      for item in doc_dict:
          if item in vocab:
            rows.append(idx)
            columns.append(vocab[item][1])
            values.append(vocab[item][0]) 
    return normalize(csr_matrix((values, (rows,columns)), shape=(len(dataset),len(vocab))))

# Printing feature names,IDF values and their respective column index in a dictionary for top 50 IDF values with feature names sorted in alphabetical order
Key is the feature and first item in list i IDF value and second item is its position in feature bag.

In [ ]:
idf_dict = fit_top_fifty(corpus)
print(idf_dict)

{'aimless': [6.922918, 0], 'artiness': [6.922918, 1], 'attempting': [6.922918, 2], 'aye': [6.922918, 3], 'baby': [6.922918, 4], 'buffet': [6.922918, 5], 'conception': [6.922918, 6], 'constructed': [6.922918, 7], 'content': [6.922918, 8], 'distressed': [6.922918, 9], 'doomed': [6.922918, 10], 'dozen': [6.922918, 11], 'drifting': [6.922918, 12], 'effort': [6.922918, 13], 'emptiness': [6.922918, 14], 'existent': [6.922918, 15], 'fill': [6.922918, 16], 'fit': [6.922918, 17], 'florida': [6.922918, 18], 'gerardo': [6.922918, 19], 'graphics': [6.922918, 20], 'highest': [6.922918, 21], 'insane': [6.922918, 22], 'massive': [6.922918, 23], 'messages': [6.922918, 24], 'muppets': [6.922918, 25], 'nearly': [6.922918, 26], 'number': [6.922918, 27], 'occurs': [6.922918, 28], 'overdue': [6.922918, 29], 'owls': [6.922918, 30], 'person': [6.922918, 31], 'post': [6.922918, 32], 'practically': [6.922918, 33], 'properly': [6.922918, 34], 'pulls': [6.922918, 35], 'punches': [6.922918, 36], 'puzzle': [6.9229

# Tesing transform function and priting idf values for Top 50 Case

In [ ]:
transform_op = transform_top_fifty(corpus,idf_dict)
print(transform_op[0].toarray())
print(len(transform_op[0].toarray()[0]))
transform_op.shape

[[0.57735027 0.         0.         0.         0.         0.
  0.         0.         0.         0.57735027 0.         0.
  0.57735027 0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.        ]]
50


(746, 50)